### PR TITLE
Add passport facebook. Big refactor

### PR DESCRIPTION
**Pull request by Survikrowa**

### CODE REVIEW

File: src/controller/codeValidationController.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import { findCodeInDb } from "../services/databaseServices/findCodeInDb";
+
+export const codeValidationController: RequestHandler = async (
+  req,
+  res,
+  next
+) => {
+  const { activationCode } = req.params;
+  const foundCode = await findCodeInDb(activationCode);
+  if (foundCode) {
```
Comment (kbkk) [inline src/controller/codeValidationController.ts:12]: I think you should "return early" :)

so refactoring the first bit of this code:
```ts
  const foundCode = await findCodeInDb(activationCode);
  if (!foundCode) {
    return res.status(400).json({ message: "Provided auth code is wrong" });
  }
  
  // ... and so on
```

File: src/services/generateCrypto.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import crypto from "crypto";
+
+export const generateCrypto: authCode = async () => {
```
Comment (kbkk) [inline src/services/generateCrypto.ts:3]: inadequate func name, perhaps `generateAuthCode` would work better

File: src/services/generateCrypto.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+  return crypto.randomBytes(16).toString("hex");
+};
+
+type authCode = () => Promise<string>;
```
Comment (kbkk) [inline src/services/generateCrypto.ts:7]: To me an `authCode` is just a `string` - maybe the generating function would return a `Promise<authCode>` but the type `authCode` itself shouldn't be a Promise

Comment (kbkk) [reply src/services/generateCrypto.ts:7]: btw. this doesn't really do anything because you'd need a "branded type" to make TS differentiate between a string and an authcode here

File: src/services/passportStrategies/passportFacebookStrategy.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+const verifyCallback: VerifyFunction = async (
+  _accessToken,
+  _refreshToken,
+  profile,
+  done
+) => {
+  if (profile?.emails) {
+    const username = profile.displayName;
+    const email = profile?.emails[0].value;
```
Comment (kbkk) [inline src/services/passportStrategies/passportFacebookStrategy.ts:24]: is it guaranteed the `emails` array will always contain at least one entry?

File: src/services/passportStrategies/passportFacebookStrategy.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    const user = await findUserBy({ email });
+    if (user) return done(null, user);
+    const crypto = await generateCrypto();
```
Comment (kbkk) [inline src/services/passportStrategies/passportFacebookStrategy.ts:27]: `generateCrypto` returns an `authCode` yet it's used to generate the initial user's password

File: src/services/userValidation.ts [hunk 6]
```diff
@@ -1,8 +1,63 @@
...
+  email,
+}: registerData) => {
+  try {
+    registerSchema.parse({ username, password, email });
+    const usernameFound = await findUserBy({ username });
+    const emailFound = await findUserBy({ email });
+    if (usernameFound || emailFound) {
+      throw [{ message: "Username or email already taken" }];
+    }
+    return true;
```
Comment (kbkk) [inline src/services/userValidation.ts:35]: to me this function has an AWFUL API.

To check if the validation succedeed I'd need to compare to true explicitly: `validateRegister(...) === true`

File: src/services/userValidation.ts [hunk 7]
```diff
@@ -1,8 +1,63 @@
...
+    } else {
+      return e;
+    }
+  }
+};
+
+export const validateLoginData = ({ username, password }: loginData) => {
+  try {
+    loginSchema.parse({ username, password });
+    return true;
```
Comment (kbkk) [inline src/services/userValidation.ts:48]: same as above

File: src/services/passportStrategies/passportLocalStrategy.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+import { Strategy as LocalStrategy } from "passport-local";
+import bcrypt from "bcrypt";
+import { validateLoginData } from "../userValidation";
+import type { ZodError } from "zod";
+import { findUserBy } from "../databaseServices/findUserBy";
+
+export const strategy = new LocalStrategy(async (username, password, done) => {
+  try {
+    const validatedSchema = validateLoginData({ username, password });
+    if (!validatedSchema[0]) {
```
Comment (kbkk) [inline src/services/passportStrategies/passportLocalStrategy.ts:10]: return early

File: src/services/databaseServices/createFacebookUser.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  password,
+  email,
+}: Partial<User>) => {
+  try {
+    return await User.create({
+      username,
+      password,
+      email,
+    });
+  } catch (error) {
```
Comment (kbkk) [inline src/services/databaseServices/createFacebookUser.ts:15]: Why do you swallow errors? If `create` fails this will usually be because of unique constraint violations or connection errors - I think it's a valid reason to throw and show return a 500 to the endpoint consumer.

Otherwise the endpoint may return a success response even though the user was not created (yikes)

File: src/services/databaseServices/createUser.ts [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+  authCodeId?: number
+) => {
+  try {
+    return await User.create({
+      username,
+      password,
+      email,
+      authCodeId,
+    });
+  } catch (error) {
```
Comment (kbkk) [inline src/services/databaseServices/createUser.ts:15]: same as above

File: src/services/databaseServices/findUserBy.ts [hunk 11]
```diff
@@ -0,0 +1,13 @@
+import { logger } from "../errorLogger";
+import { User } from "../../models/user";
+
+export const findUserBy = async (conditions: Record<string, string>) => {
+  try {
+    return User.findOne({
+      where: conditions,
+    });
+  } catch (error) {
```
Comment (kbkk) [inline src/services/databaseServices/findUserBy.ts:9]: `findOne` will return null if a row is not found, no need for the catch clause

**Review by kbkk — CHANGES_REQUESTED** [review]

File: src/models/user.ts [hunk 12]
```diff
@@ -63,80 +61,20 @@ User.init(
...
-  } catch (error) {
-    return error;
-  }
-};
-
 export const checkDB = async () => {
   try {
     await sequelize.authenticate();
     console.log("Success db connection");
   } catch (error) {
```
Comment (kbkk) [inline src/models/user.ts:76]: IMO this should rethrow so the app crashes

File: src/controller/codeValidationController.ts [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    const authCodeId = foundCode.get("authCodeId");
+    if (authCodeId) {
+      const userUpdateStatus = await updateUserAuthStatus(authCodeId);
```
Comment (nanoDW) [inline src/controller/codeValidationController.ts:15]: userUpdateStatus does not sounds like a result of updateUserAuthStatus function

File: src/controller/userController.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+import { validateRegister } from "../services/userValidation";
+import type { ZodError } from "zod";
+import { insertActivationCode } from "../services/databaseServices/insertActivationCode";
+
+export const userRegisterController: RequestHandler = async (
+  req,
+  res,
+  _next
+) => {
+  const validatedSchema = await validateRegister(req.body);
```
Comment (nanoDW) [inline src/controller/userController.ts:14]: why validatedSchema is a result of validateRegister?

File: src/controller/userController.ts [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+  if (!validatedSchema[0]) {
+    const authCode = await generateCrypto();
+    const instanceOfInsert = await insertActivationCode(authCode);
+    if (instanceOfInsert) {
+      const authCodeId = instanceOfInsert.get("authCodeId");
+      const userAccountCreateMessage = await createNewUser(req, authCodeId);
```
Comment (nanoDW) [inline src/controller/userController.ts:20]: maybe `message` could be good enough?

File: src/controller/userController.ts [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+  } else {
+    const errors = validatedSchema.map((error: ZodError) => error.message);
+    res.status(403).send({ errors });
+  }
+};
+export const userLogoutController: RequestHandler = (req, res, _next) => {
+  req.logout();
+  req.session = undefined;
+  res.clearCookie("connect.sid");
+  res.status(200).json({ message: "Successful logout" });
```
Comment (nanoDW) [inline src/controller/userController.ts:33]: do you really need a message here? maybe 204 could be a better solution?

File: src/services/databaseServices/findCodeInDb.ts [hunk 17]
```diff
@@ -0,0 +1,15 @@
+import { logger } from "../errorLogger";
+import { ActivationCodes } from "../../models/activationCode";
+
+export const findCodeInDb = async (activationCode: string) => {
```
Comment (nanoDW) [inline src/services/databaseServices/findCodeInDb.ts:4]: why async if there is no await? 👀

File: src/controller/userController.ts [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+  res,
+  _next
+) => {
+  const { username, password, email } = req.body;
+  const userCredentials = await parseUserRegisterCredentials({
+    username,
+    password,
+    email,
+  });
+  if (userCredentials[0]) {
```
Comment (kbkk) [inline src/controller/userController.ts:21]: Awful API of `parseUserRegisterCredentials`...
it either returns validation errors or the input params (?)

Honestly, if you really want to avoid throwing the error I see it this way (introducing an object holding the parsing result):
```
const parseResult = parseUserRegisterCredentials(...);
if(parseResult.hasErrors()) {
  return res.status(403).json({ errors: validationResult.errors });
}

const { username, password, email } = parseResult.data;
```

File: src/controller/userController.ts [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+    console.log(userCredentials);
+    const errors = userCredentials.map((error: ZodError) => error.message);
+    res.status(403).json({ errors });
+  } else {
```
Comment (kbkk) [inline src/controller/userController.ts:25]: return early, it's hard to follow the code

File: src/services/databaseServices/createUser.ts [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+) => {
+  try {
+    return User.create({
+      username,
+      password,
+      email,
+      authCodeId,
+    });
+  } catch (error) {
+    throw new Error(error);
```
Comment (kbkk) [inline src/services/databaseServices/createUser.ts:15]: Why do you wrap the error in another error? Just remove the catch clause

File: src/services/databaseServices/findCodeInDb.ts [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+export const findCodeInDb = (activationCode: string) => {
+  try {
+    return ActivationCodes.findOne({
+      where: {
+        activationCode,
+      },
+    });
+  } catch (error) {
+    logger.log({ level: "error", message: error });
+    return;
```
Comment (kbkk) [inline src/services/databaseServices/findCodeInDb.ts:13]: same as above

File: src/services/databaseServices/insertActivationCode.ts [hunk 22]
```diff
@@ -0,0 +1,13 @@
+import { logger } from "../errorLogger";
+import { ActivationCodes } from "../../models/activationCode";
+
+export const insertActivationCode = (activationCode: string) => {
+  try {
+    return ActivationCodes.create({
+      activationCode,
+    });
+  } catch (error) {
+    logger.log({ level: "error", message: error });
```
Comment (kbkk) [inline src/services/databaseServices/insertActivationCode.ts:10]: same as above (applies to all db functions!!)

File: src/services/generateCrypto.ts [hunk 23]
```diff
@@ -1,7 +1,5 @@
 import crypto from "crypto";
 
-export const generateAuthCode: authCode = () => {
+export const generateCrypto = async () => {
```
Comment (kbkk) [inline src/services/generateCrypto.ts:3]: why async?

**Review by kbkk — CHANGES_REQUESTED** [review]
